contracts/0.8.9/StakingRouter.sol

```diff
@@ -0,0 +1,519 @@
+        uint16 moduleFee;
+        /// @notice treasury fee
+        uint16 treasuryFee;
+        /// @notice target percent of total keys in protocol, in BP
+        uint16 targetShare;
+        /// @notice flag if module can not accept the deposits
+        bool paused;
+        /// @notice flag if module can participate in further reward distribution
+        bool active;
+        /// @notice block.timestamp of the last deposit of the module
+        uint64 lastDepositAt;
+        /// @notice block.number of the last deposit of the module
+        uint256 lastDepositBlock;
+    }
+
+    struct StakingModuleCache {
+        bool paused;
+        uint16 targetShare;
+        uint256 totalKeysCount;
+        uint256 usedKeysCount;
+        uint256 stoppedKeysCount;
+        uint256 activeKeysCount;
+        uint256 availableKeysCount;
+    }
+
+    ILido public immutable LIDO;
+
+    bytes32 public constant MANAGE_WITHDRAWAL_KEY_ROLE = keccak256("MANAGE_WITHDRAWAL_KEY_ROLE");
+    bytes32 public constant MODULE_PAUSE_ROLE = keccak256("MODULE_PAUSE_ROLE");
+    bytes32 public constant MODULE_CONTROL_ROLE = keccak256("MODULE_CONTROL_ROLE");
+    bytes32 public constant DEPOSIT_ROLE = keccak256("DEPOSIT_ROLE");
+
+    /// Version of the initialized contract data
+    /// NB: Contract versioning starts from 1.
+    /// The version stored in CONTRACT_VERSION_POSITION equals to
+    /// - 0 right after deployment when no initializer is invoked yet
+    /// - N after calling initialize() during deployment from scratch, where N is the current contract version
+    /// - N after upgrading contract from the previous version (after calling finalize_vN())
+    bytes32 internal constant CONTRACT_VERSION_POSITION = keccak256("lido.StakingRouter.contractVersion");
+
+    /// @dev Credentials which allows the DAO to withdraw Ether on the 2.0 side
+    bytes32 internal constant WITHDRAWAL_CREDENTIALS_POSITION = keccak256("lido.StakingRouter.withdrawalCredentials");
+
+    uint256 public constant DEPOSIT_SIZE = 32 ether;
+
+    uint256 internal constant TOTAL_BASIS_POINTS = 10000;
+
+    /// @dev total count of staking modules
+    uint256 private _stakingModulesCount;
+
+    /// @dev id of the last added staking module. This counter grow on staking modules adding
+    uint24 private _lastStakingModuleId;
+
+    /// @dev mapping is used instead of array to allow to extend the StakingModule
+    mapping(uint256 => StakingModule) private _stakingModules;
+
+    /// @dev Position of the staking modules in the `_stakingModules` map, plus 1 because
+    ///      index 0 means a value is not in the set.
+    mapping(uint24 => uint256) private _stakingModuleIndicesOneBased;
+
+    constructor(address _depositContract, address _lido) BeaconChainDepositor(_depositContract) {
+        if (_lido == address(0)) revert ErrorZeroAddress("_lido");
+        LIDO = ILido(_lido);
+    }
+
+    function initialize(address _admin) external {
```

Comment:
I guess the `WITHDRAWAL_CREDENTIALS_POSITION` variable initialization might be missing here. it's not good to call `setWithdrawalCredentials`, since it may trim all unused keys if the curated module is already connected at this time

Reply:
+1